### PR TITLE
Add mistral reference to context on action execution

### PR DESCRIPTION
**Pull request by m4dcoder**
Add mistral execution ID and workflow name to the context on action
execution of the main workflow. Refactor all runners to include context
in the output of the run method. Refactor the asyn worker to use the
same context to build the query context.

### CODE REVIEW

File: st2actions/st2actions/container/base.py [hunk 1]
```diff
@@ -96,18 +96,15 @@ def _do_run(self, runner, runnertype_db, action_db, actionexec_db):
...
+            (status, result, context) = runner.run(action_params)
 
             try:
                 result = json.loads(result)
             except:
                 pass
 
-            if status not in DONE_STATES:
-                self._setup_async_query(actionexec_db.id, runnertype_db, query_context)
+            if isinstance(runner, AsyncActionRunner) and status not in DONE_STATES:
```
Comment (manasdk) [inline st2actions/st2actions/container/base.py]: These are redundant checks 

- Why would any runner which is not Async return a non- DONE state?
- What happens if a non Async runner returns a non- DONE state? Do we really care if it was a non Async runner in the first place?

File: st2actions/st2actions/container/base.py [hunk 2]
```diff
@@ -96,18 +96,15 @@ def _do_run(self, runner, runnertype_db, action_db, actionexec_db):
             runner.pre_run()
 
             LOG.debug('Performing run for runner: %s', runner)
-            if isinstance(runner, AsyncActionRunner):
-                (status, result, query_context) = runner.run(action_params)
-            else:
-                (status, result) = runner.run(action_params)
+            (status, result, context) = runner.run(action_params)
```
Comment (manasdk) [inline st2actions/st2actions/container/base.py]: Why is this interface change necessary?

Comment (manasdk) [inline st2actions/st2actions/container/base.py]: Feels like we combined context and query_context. The confusing part for me is how would a runner know if something is simply context or query_context - is there any difference?

File: st2actions/st2actions/runners/mistral/v2.py [hunk 3]
```diff
@@ -170,8 +170,15 @@ def run(self, action_parameters):
                                                        **options)
 
         status = ACTIONEXEC_STATUS_RUNNING
-        query_context = {'mistral_execution_id': str(execution.id)}
-        LOG.info('Mistral query_context is %s' % query_context)
         partial_results = {'tasks': []}
 
-        return (status, partial_results, query_context)
+        context = {
```
Comment (manasdk) [inline st2actions/st2actions/runners/mistral/v2.py]: What would happen if you were only to expand the context to be structured and make no other change? Of course the corresponding change is necessary in  `st2actions/st2actions/query/mistral/v2.py` but nothing else - right?

**Comment by manasdk** [comment]
Most comments are about trying to understand scope of the change.

**Comment by manasdk** [comment]
> Refactor all runners to include context in the output of the run method.

Why?

File: st2actions/st2actions/container/base.py (hunk 1 above)
Comment (m4dcoder) [inline st2actions/st2actions/container/base.py]: I suppose we can assume that any runner that does not return a done states after run is a async runner.

File: st2actions/st2actions/container/base.py (hunk 2 above)
Comment (m4dcoder) [inline st2actions/st2actions/container/base.py]: The context contains the reference information for the mistral execution.  The query context is build by the mistral query module anyway and the context already contains the information that it needs to build the query string.  Now I can add context and query_context.  But I think it's even more confusing.  The query_context is not adding any new/different information.  So why not just delegate the runner to return appropriate context information and then let the query module build the query context it needs to do the query.  Also, I personally do not like the style that plugin can return different number of items in the tuple.

File: st2actions/st2actions/runners/mistral/v2.py (hunk 3 above)
Comment (m4dcoder) [inline st2actions/st2actions/runners/mistral/v2.py]: I'm not following.  Please rephrase.

**Comment by m4dcoder** [comment]
> Refactor all runners to include context in the output of the run method.
>> Why?

How else should I return the context information from mistral?  All the runners do not currently write to the action exec DB object directly.  The runner just returns state and result and let the container do the update.  So, to return the context information, I need to include it in the output of the run method.  So this means the other runners needs to include that as well to be consistent.  Also, I wouldn't put it in the query  context.  The query context if we need to have one is a optional function.  What if we do the state update implementation completely different?  Whereas the mistral execution context is required for the mistral runner.

**Comment by manasdk** [comment]
> How else should I return the context information from mistral?

A context was already being returned from the mistral runner. Am I missing something?

**Comment by m4dcoder** [comment]
It's returning a query context.  And I explained why I wouldn't want to put it in the query context.

**Comment by manasdk** [comment]
> What if we do the state update implementation completely different? Whereas the mistral execution context is required for the mistral runner.

I guess you disagree with the design. I get that part and understand you have reservations. Maybe we should explicitly talk through that and nail down a design or agree to the existing design and move on.

**Comment by manasdk** [comment]
So instead of output of a runner being -

`(status, result, context, query_context)`

we choose

`(status, result, context)`

Essentially context = free_form_context + query_context.

File: st2actions/st2actions/container/base.py (hunk 2 above)
Comment (manasdk) [inline st2actions/st2actions/container/base.py]: I guess you are saying context = context + query_context.

**Comment by m4dcoder** [comment]
Great we are converging.  The only question is then can other runners other than mistral return other context information (not necessary query context)?  If so, why have the runners return different number of items in the tuple.

File: st2actions/st2actions/runners/mistral/v2.py (hunk 3 above)
Comment (manasdk) [inline st2actions/st2actions/runners/mistral/v2.py]: It is irrelevant per the rest of the comments. I was trying to see how to limit scope of this PR.

A single commit often times makes it hard to follow reasoning. Also, part of your comments helped me at least understand your reasoning.

**Comment by manasdk** [comment]
> Great we are converging.

yes. But not in complete agreement yet.

> only question is then can other runners other than mistral return other context information (not necessary query context)? If so, why have the runners return different number of items in the tuple.

Well runners should return a structured object if you ask my opinion instead of this weird tuple. That will help solidify the interface. At this point if someone were to write a new runner the interface is extremely unclear. Arguably it always was. 

Anyway, I do not ask you to make the change in this PR. Clearly, you have a preference of achieving the ends here and I have no objection.

**Comment by m4dcoder** [comment]
> Well runners should return a structured object if you ask my opinion instead of this weird tuple.
Amen.

File: st2actions/st2actions/container/base.py (hunk 1 above)
Comment (manasdk) [inline st2actions/st2actions/container/base.py]: Well I do prefer the more obvious contract of only an `AsyncRunner` can return a non-DONE state. In fact I would go on to mark it as an error if a `SyncRunner` returned a non-DONE state. But that can wait and we can chat with @lakshmi-kannan to understand some of those choices.

**Comment by manasdk** [comment]
Ok with the change but I don't believe we have still nailed this interface. We should clean it up soon and bring clarity.

I can take a stab at doing that and get feedback. Including the mistral change is necessary so please merge.

+1

**Comment by m4dcoder** [comment]
@manasdk: Thanks for reviewing.  Your input is always welcome as it helps with clarifying why we are changing what we are changing.  The current implementation for runner/container is still a bit awkward.   Looking forward to it being cleaned up.

**Comment by m4dcoder** [comment]
> I guess you disagree with the design. I get that part and understand you have reservations. Maybe we should explicitly talk through that and nail down a design or agree to the existing design and move on.

We're lacking a 2nd use case for async runner.  If we have another type of async runner, I think it's easier to create a cleaner and reusable design.

**Comment by manasdk** [comment]
> We're lacking a 2nd use case for async runner.

You mean apart for Mistral? I guess ActionChainRunner also fits and should be Async. Actually we can bring Async control to a per action level and that is where I think it really resides.

**Comment by m4dcoder** [comment]
Yep, aside from Mistral.  Action chain runner is another good use case. I think if there's a 3rd party runner in the mix, it'll confirm our design.
